### PR TITLE
Fix nested pattern code generation

### LINKED ISSUE #1804: Record pattern matching with `null` gives wrong result
Spotted by Jay in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1517#issuecomment-1859821498

Given the following code:
```java
public class BooleanVariable {
	record Paper(int color) {}
	record Box<T>(T a) {}
	public static void main(String[] args) {
		Box<?> b = new Box<>(null);
		boolean res = b instanceof Box(Paper a);
		if (res) {
			System.out.println("res is true");
		} else {
			System.out.println("res is false");
		}
	}
}
```

ECJ prints "res is true" while javac prints "res is false".

### PR DESCRIPTION
**Pull request by datho7561**
## What it does
- Fixes boolean assignments that use instanceof with nested patterns.
- Fixes cases labels that use nested patterns.

Fixes #1804

## How to test

See the sample snippets that reproduce the error in the linked issue.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.compiler.batch/scripts/GenerateParserScript.class
Comment (iloveeclipse) [inline org.eclipse.jdt.core.compiler.batch/scripts/GenerateParserScript.class:1]: Please remove generated files

File: org.eclipse.jdt.core/jdtCompilerAdapter.jar
Comment (iloveeclipse) [inline org.eclipse.jdt.core/jdtCompilerAdapter.jar:1]: Is this file really meant to be changed & committed?

**Comment by danthe1st** [comment]
It might be a good idea to also create a test for something _like_ the following:
```java
public class Test{
    record Box<T>(T t){}
  
    static <T> void test(Box<T> box){
        boolean b = box instanceof Box(T t);
        System.out.println(b);
    }
  
    public static void main(String[] args){
        test(new Box<>(null));//true with javac
    }
}
```

**Comment by danthe1st** [comment]
However, the static type seems to make a difference here:
```java
public class Test{
    record Box<T>(T t){}
  
    public static void main(String[] args){
        Box<Object> box = new Box<>(null);
        switch(box){//Object
        	case Box(String s) -> System.out.println("String");
            case Box(Object s) -> System.out.println("Object");
        }
        Box<String> box2 = new Box<>(null);
        switch(box2){//String
        	case Box(String s) -> System.out.println("String");
            case Box(Object s) -> System.out.println("Object");
        }
    }
}
```

**Comment by datho7561** [comment]
@iloveeclipse @jarthana if you have time, could you please review this PR?

**Comment by mpalat** [comment]
@datho7561 I see that this pr fixes the #1796  issue as well when I checked. You can include that test case also here. Thanks!

**Comment by datho7561** [comment]
> @datho7561 I see that this pr fixes the #1796 issue as well when I checked. You can include that test case also here. Thanks!

This PR doesn't seem to fix that bug from my testing.

**Comment by srikanth-sankaran** [comment]
@datho7561 - This is my plan for this PR - I have assigned myself as a joint owner of this PR. As mentioned in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1517#issuecomment-1911691090 I am investigating and possibly undertaking a overhaul of code generation for patterns. If you don't mind, I would subsume this PR as a part of the larger work. Let us not rush this please. What do you say ?

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 1]
```diff
@@ -117,15 +117,30 @@ public void generateCode(BlockScope currentScope, CodeStream codeStream, boolean
...
+		this.pattern.fullWrapupGeneration(codeStream);
+		codeStream.removeVariable(this.secretInstanceOfPatternExpressionValue);
+		trueLabel.place();
+		codeStream.iconst_1();
+		codeStream.goto_(continueLabel);
+		falseLabel.place();
+		codeStream.iconst_0();
+		continueLabel.place();
+	} else if (this.elementVariable != null) {
 		BranchLabel actionLabel = new BranchLabel(codeStream);
```
Comment (srikanth-sankaran) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:144]: This creates dead code - `this.elementVariable` **will be** `null` when(this.pattern == null)`

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/GuardedPattern.java [hunk 2]
```diff
@@ -209,6 +209,10 @@ public void wrapupGeneration(CodeStream codeStream) {
 		this.primaryPattern.wrapupGeneration(codeStream);
 	}
 	@Override
+	public void fullWrapupGeneration(CodeStream codeStream) {
+		this.primaryPattern.fullWrapupGeneration(codeStream);
+	}
+	@Override
 	protected void generatePatternVariable(BlockScope currentScope, CodeStream codeStream, BranchLabel trueLabel,
```
Comment (srikanth-sankaran) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/GuardedPattern.java:216]: Much water has flown under the bridge - the method wrapupGeneration doesn't exist anymore on master and it doesn't likely make sense to introduce this.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 3]
```diff
@@ -117,15 +117,30 @@ public void generateCode(BlockScope currentScope, CodeStream codeStream, boolean
 
 	int pc = codeStream.position;
 
-	if (this.elementVariable != null) {
+	if (this.elementVariable != null || this.pattern != null) {
 		addAssignment(currentScope, codeStream, this.secretInstanceOfPatternExpressionValue);
 		codeStream.load(this.secretInstanceOfPatternExpressionValue);
 	} else {
 		this.expression.generateCode(currentScope, codeStream, true);
```
Comment (srikanth-sankaran) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:124]: you can simply say `if (this.pattern != null)`

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/TypePattern.java [hunk 4]
```diff
@@ -256,7 +263,7 @@ private static LocalVariableBinding getNewLocalVariableBinding(BlockScope scope,
 //		int delta =  ((TypeBinding.equalsEquals(type, TypeBinding.LONG)) ||
 //				(TypeBinding.equalsEquals(type, TypeBinding.DOUBLE))) ?
 //				2 : 1;
-		l.resolvedPosition = scope.localIndex;
+		l.resolvedPosition = scope.startIndex + scope.localIndex;
 		return l;
```
Comment (srikanth-sankaran) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/TypePattern.java:267]: This method is gone.

**Comment by srikanth-sankaran** [comment]
Interestingly I ran into this in another context and raised https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1985 - I will include the tests from here and see if the changes in `org.eclipse.jdt.internal.compiler.ast.InstanceOfExpression.generateCode(BlockScope, CodeStream, boolean)` are leverageable.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/GuardedPattern.java (hunk 2 above)
Comment (datho7561) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/GuardedPattern.java:216]: I've rebased this PR against your changes in https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/9 , I'll just need to replicate those changes here

**Comment by datho7561** [comment]
Looking through this code, I think it's worthwhile completely rewriting this fix, since the code has changed so much since when I wrote it. I'll make sure to leave the branch intact, since I think the code here might contain a hint towards the root cause.

**Pull request closed without merging**